grafana/grafana
# devenv: better loki environment

the current `loki` devenv-block uses `promtail` and maps certain log-directories from the host-computer into the docker container.
on a mac-computer this does not work well, because, for example, the mapped grafana-logs will not get updated "inside the docker container" unless the grafana process is stopped on the host-computer. so, what you will see in loki, is that a lot of logs arrived in the same second, and then no more logs ever.

so i went with a different approach, we use a nodejs-script to generate fake data. data, that is chosen so that many loki-queries can be tested with it.

i renamed the original `loki` devenv to `loki-promtail`, maybe it is useful for someone (it should work ok on Linux for example), or maybe we can fix that one at a later point somehow.

so, there are two commits here:
1. the first one renames `loki` to `loki-promtail`
2. the second adds the new `loki` devenv block

how to test:
1. `make devenv sources=loki`
2. run some loki queries. some examples:
    - `avg_over_time({place="moon"}|json float="float" | unwrap float[10s])`
        - we json-parse the log-line, extract the `float` value, and draw it's average value. there will be pauses in the graph because we intentionally add `NaN` values
    - `count_over_time({place="moon"}[1m])`
        - how many logs happen every 10 seconds. you should get a wave, because we vary the speed-of-generating-logs
    - `{place="moon"}|json|line_format "{{.counter}}"|= "150"`
        - json-parse the logs, replace the log-line with the value of the `counter` attribute, and the find a line with the value `150`
            - now you can mouse-over the log-line, and click the show-log-context button. you can easily verify that you got the correct log-row-context, because you should get counter-numbers-going-up above, and going-down below
                 - this is not perfect, because you get whole log-lines in log-row-context, also the line with `150` will be two times. these are known issues
    - `{place="moon"}|unpack`
       - we json-parse the line, and replace the line with the `_entry` attribute's value. you can see that ascii-color-codes work
3. if you need some custom data in Loki, you can just modify the nodejs-script, and run it directly with `node devenv/docker/blocks/loki/data/data.js http://localhost:3100`
